drivers/tee/optee/optee.c

```diff
@@ -434,15 +437,127 @@ static int optee_invoke_func(const struct device *dev, struct tee_invoke_func_ar
 	return (rc) ? rc : ret;
 }
 
-static int optee_shm_register(const struct device *dev, void *addr, uint64_t size, uint32_t flags,
-			      struct tee_shm *shm)
+static void *optee_construct_page_list(void *buf, uint32_t len, uint64_t *phys_buf)
 {
-	return 0;
+	const uint32_t page_size = OPTEE_MSG_NONCONTIG_PAGE_SIZE;
+	const uint32_t num_pages_in_pl = page_size - 1;
+	uint32_t page_offset = (uintptr_t)buf & (page_size - 1);
+
+	uint8_t *buf_page;
+	uint32_t num_pages;
+	uint32_t list_size;
+
+	/* see description of OPTEE_MSG_ATTR_NONCONTIG */
+	struct {
```

Comment:
You are doing a wild assumption that compiler will not put any padding in this struct.

Add `BUILD_ASSERT(sizeof(*p) == OPTEE_MSG_NONCONTIG_PAGE_SIZE)` at least

Reply:
+